src/gossip/map.zig

```diff
@@ -0,0 +1,357 @@
+        return self.gossip_data.getTypedConst(tag, self.key_to_index.values()[index]);
+    }
+
+    pub fn tagOfIndex(self: *const GossipMap, index: usize) GossipDataTag {
+        return self.key_to_index.values()[index].tag;
+    }
+
+    pub fn swapRemove(self: *GossipMap, key: GossipKey) bool {
+        const index = self.key_to_index.getIndex(key) orelse return false;
+        const tagged_index = self.key_to_index.values()[index];
+
+        self.key_to_index.swapRemoveAt(index);
+        _ = self.gossip_data.swapRemove(tagged_index);
+        _ = self.metadata.swapRemove(index);
+
+        // reindex the moved item, if there was one
+        const tag_len = self.gossip_data.tagLen(tagged_index.tag);
+        assert(tagged_index.index <= tag_len);
+        if (tagged_index.index < tag_len) {
+            const newly_swapped = self.gossip_data.get(tagged_index);
+            const index_entry = self.key_to_index.getEntry(newly_swapped.label()).?;
+            index_entry.value_ptr.* = tagged_index;
+        }
+
+        return true;
+    }
+
+    pub fn getOrPut(self: *GossipMap, allocator: Allocator, key: GossipKey) !GetOrPut {
+        const key_gop = try self.key_to_index.getOrPut(allocator, key);
+
+        const metadata_ptr, const list_entry = if (key_gop.found_existing) .{
+            &self.metadata.items[key_gop.index], self.gossip_data.getEntry(key_gop.value_ptr.*), //
+        } else blk: {
+            const metadata_ptr = try self.metadata.addOne(allocator);
+            errdefer _ = self.metadata.pop();
+            assert(key_gop.index == self.metadata.items.len - 1);
+
+            const list_entry = try self.gossip_data.addOne(allocator, key);
+            key_gop.value_ptr.* = list_entry.index;
+
+            break :blk .{ metadata_ptr, list_entry };
+        };
+
+        return .{
+            .entry = .{
+                .key_ptr = key_gop.key_ptr,
+                .metadata_ptr = metadata_ptr,
+                .gossip_data_entry = list_entry,
+                .index = key_gop.index,
+            },
+            .found_existing = key_gop.found_existing,
+        };
+    }
+
+    pub const GetOrPut = struct {
+        entry: Entry,
+        found_existing: bool,
+    };
+
+    pub fn iterator(self: *const GossipMap) Iterator {
+        const slice = self.key_to_index.entries.slice();
+        return .{
+            .keys = slice.items(.key),
+            .indices = slice.items(.value),
+            .gossip_data = &self.gossip_data,
+            .metadata = self.metadata.items,
+        };
+    }
+
+    pub const Iterator = struct {
+        keys: []GossipKey,
+        indices: []SplitUnionList(GossipData).Index,
+        gossip_data: *const SplitUnionList(GossipData),
+        metadata: []Metadata,
+        cursor: usize = 0,
+
+        pub fn next(self: *Iterator) ?Entry {
+            defer self.cursor += 1;
+            return if (self.cursor < self.keys.len) .{
+                .key_ptr = &self.keys[self.cursor],
```

Comment:
why not just give iterator `self` and then call `getEntryByIndex` ?

Reply:
That would be more computationally expensive, though it may be a negligible difference.